apps/wolfsshd/wolfsshd.c

```diff
@@ -1771,81 +1829,59 @@ static void* HandleConnection(void* arg)
-                        SHELL_Subsystem(conn, ssh, pPasswd, usrConf,
-                                wolfSSH_GetSessionCommand(ssh));
-                    }
-                #endif /* WOLFSSH_SHELL */
-
-                    /* SCP can be an exec type */
-                    if (ret == WS_SCP_INIT) {
-                    #ifdef WOLFSSH_SCP
-                        ret = SCP_Subsystem(conn, ssh, pPasswd, usrConf);
-                    #else
-                        err_sys("SCP not compiled in. Please use "
-                                "--enable-scp");
-                    #endif
-                    }
-                    break;
+        do {
+            ret = wolfSSH_worker(ssh, NULL);
+        } while (conn->channel == NULL);
 
-                case WOLFSSH_SESSION_TERMINAL:
-                default:
-                    wolfSSH_Log(WS_LOG_ERROR,
-                        "[SSHD] Unknown or build not supporting session type "
-                        "found");
-                    ret = WS_NOT_COMPILED;
+        if (conn->doShell) {
+        #ifdef WOLFSSH_SHELL
+            if (ret == WS_SUCCESS) {
+                SHELL_Subsystem(conn, ssh, pPasswd, usrConf, NULL);
+            }
+        #else
+            wolfSSH_Log(WS_LOG_ERROR, "[SSHD] Shell support is disabled");
+            ret = WS_NOT_COMPILED;
+        #endif
+        }
+        else if (conn->doSftp) {
+        #ifdef WOLFSSH_SFTP
+            int select_ret;
+            do {
+                select_ret = tcp_select(conn->fd, 1);
+                if (select_ret == WS_SELECT_RECV_READY
+                        || select_ret == WS_SELECT_ERROR_READY
+                        || error == WS_WANT_WRITE) {
+                    ret = wolfSSH_SFTP_accept(ssh);
+                    error = wolfSSH_get_error(ssh);
+                }
+                else if (select_ret == WS_SELECT_TIMEOUT) {
+                    error = WS_WANT_READ;
+                }
+                else {
+                    error = WS_FATAL_ERROR;
+                }
+            } while (ret != WS_SFTP_COMPLETE
+                    && (error == WS_WANT_READ || error == WS_WANT_WRITE));
+            if (ret == WS_SFTP_COMPLETE) {
+                ret = SFTP_Subsystem(conn, ssh, pPasswd, usrConf);
+            }
+        #else
+            err_sys("SFTP not compiled in. Please use --enable-sftp");
+        #endif
+        }
+        else if (conn->doScp) {
+        #ifdef WOLFSSH_SCP
+            ret = SCP_Subsystem(conn, ssh, pPasswd, usrConf);
+        #else
+            err_sys("SCP not compiled in. Please use --enable-scp");
+        #endif
+        }
+        else if (conn->doExec) {
+            if (ret == WS_SUCCESS) {
+                wolfSSH_Log(WS_LOG_INFO, "[SSHD] Entering exec session [%s]",
+                    wolfSSH_GetSessionCommand(ssh));
+                SHELL_Subsystem(conn, ssh, pPasswd, usrConf,
+                        wolfSSH_GetSessionCommand(ssh));
             }
```

Comment:
would be nice to have a final else case to log that an unknown/unsupported channel type was recieved

Reply:
wolfSSHd's channel request callbacks set those flags after the channel is created, it's guaranteed. Having a channel pointer and none of the three "do" flags sets is definitely an error condition. Adding the else clause.